here-naksha-app-service/src/test/java/com/here/naksha/app/service/WriteFeaturesAtomicityTest.java

```diff
@@ -0,0 +1,129 @@
+    String createFeaturesJson = loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/create_features.json");
+
+    // When: saving these features
+    String streamId = UUID.randomUUID().toString();
+    HttpResponse<String> saveResp = nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeaturesJson, streamId);
+
+    // Then: response indicates failure
+    ResponseAssertions.assertThat(saveResp)
+        .hasStatus(500)
+        .hasStreamIdHeader(streamId);
+
+    // And: none of the features got saved
+    String idsQuery = "?id=" + String.join("&id=", validFeatureIds);
+    HttpResponse<String> getResp = nakshaClient.get("hub/spaces/" + SPACE_ID + "/features" + idsQuery, streamId);
+    ResponseAssertions.assertThat(getResp)
+        .hasStatus(200)
+        .hasStreamIdHeader(streamId)
+        .hasJsonBody(loadFileOrFail("WriteFeaturesAtomicity/TC1101_partialCreateShouldFail/empty_response.json"));
+  }
+
+  @Test
+  void tc_1102_partialUpdateShouldFail() throws URISyntaxException, IOException, InterruptedException {
+    // Given: multiple features to save (all valid)
+    String streamId = UUID.randomUUID().toString();
+    List<String> featureIds = List.of("tc_1102_feature_1", "tc_1102_feature_2");
+    String createFeatureJson = loadFileOrFail("WriteFeaturesAtomicity/TC1102_partialUpdateShouldFail/create_features.json");
+    nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeatureJson, streamId);
+
+    // When: updating these features and one of the updates is invalid (missing points coordinates)
+    HttpResponse<String> updateResp = nakshaClient.put("hub/spaces/" + SPACE_ID + "/features",
+        loadFileOrFail("WriteFeaturesAtomicity/TC1102_partialUpdateShouldFail/update_features.json"),
+        streamId);
+
+    // Then: response indicates failure
+    ResponseAssertions.assertThat(updateResp)
+        .hasStatus(500)
+        .hasStreamIdHeader(streamId);
+
+    // And: none of the features got updated - they are equal to initial (creation) state
+    String idsQuery = "?id=" + String.join("&id=", featureIds);
+    HttpResponse<String> getResp = nakshaClient.get("hub/spaces/" + SPACE_ID + "/features" + idsQuery, streamId);
+    ResponseAssertions.assertThat(getResp)
+        .hasStatus(200)
+        .hasStreamIdHeader(streamId)
+        .hasJsonBody(createFeatureJson);
+  }
+
+  @Test
+  void tc_1103_partialDeleteShouldSucceed() throws URISyntaxException, IOException, InterruptedException {
+    // Given: multiple features to save (all valid)
+    String streamId = UUID.randomUUID().toString();
+    List<String> createdFeatureIds = List.of("tc_1103_feature_1", "tc_1103_feature_2");
+    String createFeatureJson = loadFileOrFail("WriteFeaturesAtomicity/TC1103_partialDeleteShouldSucceed/create_features.json");
+    nakshaClient.post("hub/spaces/" + SPACE_ID + "/features", createFeatureJson, streamId);
```

Comment:
nth: like the setups, adding assert 200 status code here might help spot unexpected problem sooner

Reply:
will do :)